File: sample_source_code/python/src/animal_visitor.py
"""AnimalVisitor: collects aggregate stats by traversing animals."""
from __future__ import annotations

from ivisitor import IVisitor
from math_utils import MathUtils


class AnimalVisitor(IVisitor):
    """
    Concrete visitor that accumulates per-species counts and health scores.
    Demonstrates double-dispatch via visit_dog / visit_cat.
    """

    def __init__(self) -> None:
        self._dog_count    = 0
        self._cat_count    = 0
        self._total_health = 0.0

    def visit_dog(self, dog) -> None:
        self._dog_count    += 1
        self._total_health += dog.compute_health_score()

    def visit_cat(self, cat) -> None:
        self._cat_count += 1
        proxy = MathUtils.clamp(100.0 - abs(cat.weight - 4.5) * 10, 0.0, 100.0)
        self._total_health += proxy

    @property
    def dog_count(self) -> int:   return self._dog_count

    @property
    def cat_count(self) -> int:   return self._cat_count

    @property
    def total_count(self) -> int: return self._dog_count + self._cat_count

    def average_health_score(self) -> float:
        if self.total_count == 0:
            return 0.0
        return round(self._total_health / self.total_count, 2)

    def reset(self) -> None:
        self._dog_count    = 0
        self._cat_count    = 0
        self._total_health = 0.0

File: sample_source_code/javascript/src/AnimalVisitor.js
const IVisitor  = require('./IVisitor');
const MathUtils = require('./MathUtils');

/**
 * Concrete visitor: accumulates health stats while traversing animals.
 * @extends IVisitor
 */
class AnimalVisitor extends IVisitor {
    constructor() {
        super();
        this._dogCount    = 0;
        this._catCount    = 0;
        this._totalHealth = 0;
    }

    visitDog(dog) {
        this._dogCount++;
        this._totalHealth += dog.computeHealthScore();
    }

    visitCat(cat) {
        this._catCount++;
        const proxy = MathUtils.clamp(100 - Math.abs(cat.getWeight() - 4.5) * 10, 0, 100);
        this._totalHealth += proxy;
    }

    get dogCount()   { return this._dogCount; }
    get catCount()   { return this._catCount; }
    get totalCount() { return this._dogCount + this._catCount; }

    averageHealthScore() {
        return this.totalCount === 0
            ? 0
            : Math.round((this._totalHealth / this.totalCount) * 100) / 100;
    }
}

module.exports = AnimalVisitor;
